[IMP] event: allow multiple entries for tickets

frva-odoo opened the pull request:
This PR adds multi-entry on the event registration model, meaning that one bought ticket could be used multiple times to enter an event.

Current behavior before PR:

There are only single use tickets.

Desired behavior after PR is merged:

The registration should be ongoing as long as there is remaining entries if there is multiple entries allowed on the ticket used by the registration, and is usable to enter the event.
Single use ticket are not affected by those changes and their behavior is the same as before.

Task-5373441

vava-odoo commented on addons/event/models/__init__.py:
```diff
@@ -11,6 +11,7 @@
 from . import event_mail_registration
 from . import event_mail_slot
+from . import event_registration_attendance
 from . import event_registration
```
```suggestion
from . import event_registration
from . import event_registration_attendance
```

vava-odoo commented on addons/event/models/event_registration.py:
```diff
@@ -312,6 +352,30 @@ def action_set_draft(self):
+                    'registration_id': record.id,
+                    'attendance_date': datetime.today(),
+                })
+                record.attendance_last_lines_amount = len(record.attendances_ids)
```
You could make a compute method for this field instead of duplicating this everywhere you modify `attendances_ids`

vava-odoo commented on addons/event/models/event_registration.py:
```diff
@@ -312,6 +352,30 @@ def action_set_draft(self):
+                else:
+                    record.action_set_done()
+            else:
+                record.action_set_done()
```
There should be a way not to call 3 times `record.action_set_done()`. Isn't it equivalent to this (`continue` and skip the call only in this specific case)?
```suggestion
                    if record.remaining_entries > 0:
                        continue
            record.action_set_done()
```

vava-odoo commented on addons/event/views/event_registration_views.xml:
```diff
@@ -116,6 +120,14 @@
                                 </kanban>
                             </field>
                         </page>
+                        <page string="Attendances" name="attendances" invisible="False">
```
```suggestion
                        <page string="Attendances" name="attendances">
```

vava-odoo commented on addons/event/models/event_registration.py:
```diff
@@ -210,6 +225,13 @@ def _check_event_ticket(self):
+    def _constrains_dates_coherency(self):
+        for registration in self:
+            if registration.remaining_entries < 0:
+                raise UserError(_('The remaining entries cannot be lower than 0. '
```
```suggestion
                raise UserError(self.env._('The remaining entries cannot be lower than 0. '
```

vava-odoo commented on addons/event/models/event_registration.py:
```diff
@@ -312,6 +354,27 @@ def action_set_draft(self):
 
+    def cancel_last_attendance(self):
+        for record in self:
+            changes = {'attendances_ids': [(2, self.attendances_ids[-1].id)], 'attendance_last_lines_amount': record.attendance_last_lines_amount - 1}
```
```suggestion
            changes = {
                'attendances_ids': [(2, self.attendances_ids[-1].id)],
                'attendance_last_lines_amount': record.attendance_last_lines_amount - 1,
            }
```

tde-banana-odoo commented on addons/event/models/event_registration_attendance.py:
```diff
@@ -0,0 +1,12 @@
+from odoo import fields, models
+
+
+class EventRegistrationAttendance(models.Model):
```
Hello,

We spoke a bit about it in the team and that is a bit overengineering ... using 2many adds complexity and make the model suddenly strange.

It would be simpler to simply create multiple registrations, and maybe keep a "main_registration_id" field on sub-regisrations. It was something we planned to do at some point notably for free orders and event_crm bridge, so probably the time to do it.

Idea: have a main registration and sub registrations. When you browse attendees of a given event with multiple comings, you can see that a given partner came several times, and that's actually better (you can group by date, ...) compared to a sub model that hides the real attendees.

Also it can then be used in event_crm, as we have rules to group regisrations. Currently it is based on create_date when we don't have another grouping field (SO), can be used to improve that heuristic.

Could you update the PR accordingly ? Thanks !

jeh-odoo commented on addons/event/models/event_registration.py:
```diff
@@ -312,6 +337,30 @@ def action_set_draft(self):
+    def cancel_last_attendance(self):
+        sub_registrations = self.env['event.registration'].search([('main_registration_id', 'in', self.ids), ('state', '=', 'done')])
+        for record in self:
+            related_registrations_sorted = sub_registrations.filtered(lambda r: r.main_registration_id.id == record.id).sorted('date_closed desc')
```
see above comment to avoid that filtered

frva-odoo commented on addons/event/models/event_registration.py:
```diff
@@ -312,6 +337,30 @@ def action_set_draft(self):
+    def action_attend_event(self):
+        for record in self:
+            if record.remaining_entries > 1:
+                new_registration = record.with_context(install_mode=True).copy({
```
This is here so that the new registration skips the creation of a ticket and a mail sent to the customer. When it is done it slows down greatly the creation so we have to skip if we want to keep the scan of a ticket quick. 
This is the best way I found to do this, tell me if there is a "cleaner" way as I agree it is quite weird to simulate install_mode.

reth-odoo replied:
Couldn't you do it with a heuristic in create? i.e. if main_registration_id is set, don't update the scheduler? Or is there a case where that would be desirable?

Otherwise would probably need a new context key like `mail_notrack`
You call it simulating I call it lying to the computer 🥸

tde-banana-odoo replied:
You can use `tracking_disable` context key then, with a comment telling why you do it.

frva-odoo commented on addons/event/models/event_registration.py:
```diff
@@ -312,6 +337,30 @@ def action_set_draft(self):
+                new_registration = record.with_context(install_mode=True).copy({
+                    'main_registration_id': record.id,
+                })
+                new_registration.action_set_done()
```
This is to have all things that happens when we modify it being correctly called; for exemple we have a message logged on the registration on write, or we also have the date_closed field filled when the registration is done. Both of these things don't happen if we create it as done directly

frva-odoo commented on addons/event/views/event_registration_views.xml:
```diff
@@ -81,6 +81,10 @@
                         </group>
+                        <group string="Transaction" name="transaction">
+                            <field name="remaining_entries" invisible="not ticket_entry_limit or main_registration_id"/>
+                            <field name="main_registration_id" readonly="true" invisible="not main_registration_id"/>
```
I corrected it for this as intended, but I also had to remove the invisible condition on other modules using this group. There is still one case where it could be empty (installing event and pos_event) but as this group was already existing empty before this PR, I thought it was better to let it be in this case instead of having a possible case where the group is empty when there are fields inside (if we install event_sale then pos_event with an invisible condition in pos_event it would get it but could raise issues)

frva-odoo commented:
Hello @jeh-odoo,
Thank you very much for your review 🙂 
I corrected what I could and left a few comments for the remaining issues. 
I also added two small features asked by the PO: a filter in the registrations search, and hidding the undo button when no ticket was valided on scan.
Cheers !

reth-odoo commented on addons/event/models/event_registration.py:
```diff
@@ -191,6 +199,15 @@ def _compute_event_end_date(self):
 
+    @api.depends("ticket_entry_limit")
+    def _compute_remaining_entries(self):
+        sub_registrations = self.env['event.registration'].search([('main_registration_id', 'in', self.ids), ('state', '=', 'done')])
```
readgroup aggregating ids as recordset, grouping by main_registration_id?

reth-odoo commented on addons/event/models/event_registration.py:
```diff
@@ -233,6 +250,14 @@ def _onchange_phone_validation(self):
+            if registration.ticket_entry_limit and registration.state == 'open' and registration.remaining_entries < 1:
+                registration.action_set_done()
+            elif registration.state == 'done' and registration.remaining_entries > 0:
+                registration.action_confirm()
```
Hm.. wouldn't this make sense in `write`? It's not really specific to the UI and you don't really want to apply the change unless people save anyways

tde-banana-odoo replied:
This globally makes no sense ... remaining_entries is a computed field, does not need any onchange. Could set an inverse, but don't see how it should be reflected in compute.

Not sure what it tries to do but seems wrong to me. Maybe serves as a shortcut for a real action ? What is the purpose, what are you trying to implement as business flow ?

frva-odoo replied:
You're both right, it was here when I used attendance with the remaining entries as a stored computed field at the first steps of the task, but now it is completely irrelevant and so I can remove it.

reth-odoo commented on addons/event/models/event_registration.py:
```diff
@@ -466,4 +505,6 @@ def _get_registration_summary(self):
             'badge_format': self.event_id.badge_format,
             'date_closed_formatted': format_date(env=self.env, value=self.date_closed, date_format='short') if self.date_closed else False,
             'is_date_closed_today': is_date_closed_today,
+            'remaining_entries': max(0, self.remaining_entries - 1),
```
Would say it deserves a comment

reth-odoo commented on addons/event/models/event_ticket.py:
```diff
@@ -37,6 +37,10 @@ def default_get(self, fields):
+    entry_limit = fields.Integer(default=0,
+        help="Enable multi-entry tickets\n"
+        "- Set to 0 to disable.\n"
+        "- Enter 1 or more to define the maximum number of allowed entries.")
```
Seems a bit weird, isn't the disabled behavior already that you can only enter once? So it feels like the minimum should be 1 and 1 behaves as before.

It seems the only difference is if you pick 1 you can cancel the attendance in the UI (?)

tde-banana-odoo replied:
Would consider that 0 / False / unset or 1 means entry is limited. Anything more than 1 means you set a higher limit. That way we don't need any migration nor default (unset = old behavior).

tde-banana-odoo commented on addons/event/models/event_registration.py:
```diff
@@ -31,6 +31,10 @@ def _get_random_barcode(self):
         return str(int.from_bytes(os.urandom(8), 'little'))
 
+    @api.model
+    def _default_event_ticket_id(self):
```
Weird to me to set default ticket. Should be an active choice, not a default one ... what are you trying to solve ?

frva-odoo replied:
It was because we thought it would be weird to have registrations without a ticket hence why we put a default, but if you say it should be an active choice it is better to remove it then

tde-banana-odoo commented on addons/event/models/event_registration.py:
```diff
@@ -78,6 +83,9 @@ def _get_random_barcode(self):
              'Cancelled: registrations cancelled manually')
+    remaining_entries = fields.Integer(string="Remaining Entries", compute="_compute_remaining_entries", tracking=True)
+    ticket_entry_limit = fields.Integer(string="Initial Entry Limit", related="event_ticket_id.entry_limit", tracking=True)
+    main_registration_id = fields.Many2one('event.registration', string='Main Registration')
```
Would probably add an ondelete='restrict', forcing people to remove sub attendees first.

tde-banana-odoo commented on addons/event/views/event_ticket_views.xml:
```diff
@@ -74,7 +74,8 @@
                 <field name="name"/>
                 <field name="description" optional="hide"/>
-                <field name="start_sale_datetime" optional="hide"/>
+                <field name="entry_limit" string="Entry Limit" optional="hide"/>
```
useless string, should be defined on field if necessary

tde-banana-odoo commented on addons/event/views/event_registration_views.xml:
```diff
@@ -216,6 +220,8 @@
                 <filter string="Registered" name="confirmed" domain="[('state', '=', 'open')]"/>
                 <filter string="Attended" name="attended" domain="[('state', '=', 'done')]"/>
                 <separator/>
+                <filter string="Main Registration" name="main_registration" domain="[('main_registration_id', '=', False)]"/>
```
Shouldn't we also add a filter to allow to group_by main_registration_id ? If you use those, you will probably try to group_by to have a "hierarchical" view of registrations, isn't it ?

tde-banana-odoo commented on addons/event/models/event_registration.py:
```diff
@@ -312,6 +337,33 @@ def action_set_draft(self):
     def action_confirm(self):
         self.write({'state': 'open'})
 
+    def cancel_last_attendance(self):
```
Don't understand this and the linked JS code

frva-odoo replied:
This is for when you validate a registration through the registration desk, before closing the validation window, you can click the "undo" button. This will cancel the use of the ticket you validated (if you clicked on the wrong ticket by error) so if this was a multi-entry and there is still entries remaining (i.e. you created a new sub registration), cancel that one and otherwise put the ticket back to state=open.

tde-banana-odoo commented on addons/event/models/event_registration.py:
```diff
@@ -312,6 +337,33 @@ def action_set_draft(self):
+                sub_registration_link = new_registration._get_html_link(title=f"{new_registration.name}")
+                record._message_log(
+                    body=_(
+                        "A related sub-registration is validated to represent the attendance of one entry for this main registration: %s", sub_registration_link
```
Seems highly verbose and dev-oriented to me... maybe something more neutral like
```suggestion
                        "New attendance by this attendee created at %(attendance_datetime)s, see %(attendance_link)s",
```

tde-banana-odoo commented on addons/event/models/event_registration.py:
```diff
@@ -312,6 +337,33 @@ def action_set_draft(self):
+                    body=_(
+                        "A related sub-registration is validated to represent the attendance of one entry for this main registration: %s", sub_registration_link
+                    ),
+                    message_type='comment',
```
Keep notification type, as it is a system generated message (comment are user inputs)

reth-odoo commented on addons/event/models/event_registration.py:
```diff
@@ -191,6 +194,19 @@ def _compute_event_end_date(self):
+            aggregates=['id:recordset']
+        ))
+        for registration in self:
+            if registration.ticket_entry_limit > 1 and not registration.main_registration_id and registration.state != 'done':
+                registration.remaining_entries = (registration.ticket_entry_limit - len(sub_registrations.get(registration, [])))
```
As there's technically nothing preventing creating more than the limit
Would add a `min(0, ...)`

Also you can use the __count aggregator rather than id:recordset if we only care about that.
🐌 🌩️

reth-odoo commented on addons/event/models/event_registration.py:
```diff
@@ -312,6 +329,38 @@ def action_set_draft(self):
+        last_attendance_registrations = self.filtered(lambda r: r.remaining_entries == 0)
+        other_registrations = self.filtered(lambda r: r.remaining_entries > 0)
+        last_attendance_registrations.write({'state': 'open'})
+        last_sub_registrations = self.env['event.registration']._read_group(
+            domain=[('main_registration_id', 'in', other_registrations.ids), ('state', '=', 'done')],
+            groupby=['main_registration_id'],
+            aggregates=['date_closed:max', 'id:max']
+        )
+        self.env['event.registration'].browse(
+            [reg_id for main_reg, date, reg_id in last_sub_registrations]
+        ).action_cancel()
```
Doesn't seem we use the aggregated date_closed, maybe you meant to order by it?

frva-odoo replied:
Yes that is the goal, I want to get the id of the last validated sub-registration for each main registration concerned here. It seems to me like this is the best way because if I use order="...", this orders the lines after the groupby so I will have multiple lines for the same main_registration_id where i would still need to filter afterwards. Is there a better way to do it than how I do now ?

reth-odoo replied:
No but I mean `date_closed:max` is never actually used anywhere. The way this works, `id:max` is unrelated to `date_closed:max`.

If the idea was to use aggregates to get the same as `order="date_closed DESC, id DESC"`, you just can't do it with a standard readgroup. I don't really see a way to do it in 1 query without custom SQL.

I'd say we can keep the "last id" logic as is (we can remove date_closed without change in behavior) as anyways unless you're fiddling with stuff, the last closed in this context is going to be the one with the highest id.

--------------------

Maybe we can have a docstring to specify the expected behavior

frva-odoo replied:
Yes aggregating with id:max only does what I want as long as the date_closed are not changed, so I can do it this way. Thank you for your help on this.

reth-odoo commented on addons/event/models/event_registration.py:
```diff
@@ -312,6 +329,38 @@ def action_set_draft(self):
         self.write({'state': 'open'})
 
+    def cancel_last_attendance(self):
+        last_attendance_registrations = self.filtered(lambda r: r.remaining_entries == 0)
```
I suppose we're assuming all records in `self` are "main registrations"? I would definitely specify that in docstring, and probably raise a UserError if not the case as otherwise it breaks all the logic, since this is true of all subregistrations too.

reth-odoo commented on addons/event/models/event_registration.py:
```diff
@@ -312,6 +329,38 @@ def action_set_draft(self):
+        )
+        self.env['event.registration'].browse(
+            [reg_id for main_reg, date, reg_id in last_sub_registrations]
+        ).action_cancel()
```
We cancel the last one but then it's just going to sit there, then when you scan the ticket it will create a new one and the other one will just be dangling right?

reth-odoo replied:
Considering the use case now I suppose it's ok, it is indeed functionally cancelled

tde-banana-odoo commented:
> It seems weird that we just have a bunch of duplicate attendees hanging around. They probably should be hidden in default filters and not count towards total attendees and such

I quite disagree. If someone comes 10 times at your museum, you still have 10 entries. For OXP we should not use it to know who came on a given day, because it is a bad usage of it. The real target according to me are museum, theatres, ... where each entry counts as a real one, even if it is the same person.

reth-odoo commented:
Feel free to ignore my comments about hiding/not sending messages to "fake attendees" for now, I was unclear on the use case.

tde-banana-odoo commented:
Actually I am unsure what we are trying to do. Which generally does not help coding stuff.

reth-odoo commented on addons/event/static/src/client_action/event_registration_summary_dialog.js:
```diff
@@ -61,7 +56,7 @@ export class EventRegistrationSummaryDialog extends Component {
     async undoRegistration() {
         if (["confirmed_registration", "already_registered"].includes(this.registrationStatus.value)) {
-            await this.orm.call("event.registration", "action_confirm", [this.registration.id]);
+            await this.orm.call("event.registration", "cancel_last_attendance", [this.registration.id]);
```
We go from confirming it to canceling it?

frva-odoo replied:
This was confusing for me at first too but it is the same behavior as before: the method action_confirm was previously called because the way to "undo" a validated ticket is to put it back to the state "open", and that is what action_confirm does so that was why it was used. Now I put the cancel_last_attendance method which does exactly this for ticket with 0 remaining_entries (fully used multi-entry tickets, classic tickets) and cancel a sub registration otherwise.

reth-odoo replied:
Oh right it's "confirmed" not "attended" :thinking: 
Indeed, thanks for the explanation :)

reth-odoo commented on addons/event/models/event_registration.py:
```diff
@@ -279,7 +295,8 @@ def create(self, vals_list):
         registrations = super().create(vals_list)
-        registrations._update_mail_schedulers()
+        # Avoid sub-registrations so no mail/ticket is sent again to the client and we speed up creation
+        registrations.filtered(lambda r: not r.main_registration_id)._update_mail_schedulers()
```
Although we don't "update" the scheduler, it will still run later and notice there's a new registration that was not notified. We actually need a filter in event_mail if we want this to work reliably.

(To see whether we do want to silence new registration notifications, as per your comment on the task)

frva-odoo replied:
I was not aware they would still be sent, thank you for seeing this. I will make sure to adapt it based on what is decided about the mails/sms

reth-odoo commented on addons/event/views/event_registration_views.xml:
```diff
@@ -81,6 +81,10 @@
                             <field name="create_date" string="Registration Date" groups="base.group_no_one"/>
                             <field name="date_closed" groups="base.group_no_one"/>
                         </group>
+                        <group string="Transaction" name="transaction" invisible="ticket_entry_limit  &lt; 2">
```
Does this take up space if all the children are invisible? Otherwise maybe we don't need this (avoids needing overrides in the 2 modules that make this always-visible)

frva-odoo replied:
I checked and it does take spaces even if there's nothing inside, so it is probably better to hide it like this.

reth-odoo replied:
Indeed

reth-odoo commented on addons/event/static/src/client_action/event_registration_summary_dialog.xml:
```diff
@@ -30,10 +30,11 @@
                             </t>
                             <t t-else="">
-                                <span>Ticket already scanned!</span>
+                                <span style="font-size:16px">This Card/Ticket is currently unavailable.<br/>
+                                    The offer may have expired or you may have reached the maximum number of entries allowed.</span>
```
With multi-line it looks quite weird imo

`ms-sm-5` on the fa-exclamation-circle above isn't super relevant for such a long message.
Wouldn't spend too much time but if possible to simply align icon and text to the center and let flexbox decide what the "margin" for the icon is, would be nice.

tde-banana-odoo commented on addons/event/models/event_registration.py:
```diff
@@ -312,6 +329,48 @@ def action_set_draft(self):
+        last_sub_registrations = self.env['event.registration']._read_group(
+            domain=[('main_registration_id', 'in', other_registrations.ids), ('state', '=', 'done')],
+            groupby=['main_registration_id'],
+            aggregates=['id:max']
```
Would probably set the order to be id DESC (current default model ordering, but if it changes, this feature is broken)

frva-odoo replied:
I don't see the relevance of ordering by id here, we have a readgroup that associate one main_registration_id with the sub-registration having the highest id (being by default the last one created). So even if the model is ordered differently, the max stays the same and order of the main_registration doesn't matter. Tell me if I'm wrong here

tde-banana-odoo commented on addons/event/models/event_registration.py:
```diff
@@ -312,6 +329,48 @@ def action_set_draft(self):
+        or make it open again if there was no remaining entries.
+
+        :raises: UserError if one of the registrations is a sub-registrations, this method is meant to be used only on main registrations. """
+        if any(self.mapped("main_registration_id")):
```
Pythonic usage of recorsets
```suggestion
        if self.mapped("main_registration_id"):
```

tde-banana-odoo commented on addons/event/models/event_registration.py:
```diff
@@ -312,6 +329,48 @@ def action_set_draft(self):
+
+        :raises: UserError if one of the registrations is a sub-registrations, this method is meant to be used only on main registrations. """
+        if any(self.mapped("main_registration_id")):
+            raise UserError("You cannot cancel the last attendance of this registration.")
```
Try to improve message, as this one does not help the user understanding anything. Maybe add failing registration IDs, tell they should somehow cancel from the "master" registration, ...

tde-banana-odoo commented on addons/event/models/event_registration.py:
```diff
@@ -312,6 +329,48 @@ def action_set_draft(self):
+        if any(self.mapped("main_registration_id")):
+            raise UserError("You cannot cancel the last attendance of this registration.")
+        last_attendance_registrations = self.filtered(lambda r: r.remaining_entries == 0)
+        other_registrations = self.filtered(lambda r: r.remaining_entries > 0)
```
Don't understand why "cancel last attendance" does not cancel last attendance in all cases. Don't really see why we split on remaining entries ... seems working more by side effects than clear modeling.

tde-banana-odoo commented on addons/event/models/event_ticket.py:
```diff
@@ -49,6 +53,11 @@ def default_get(self, fields):
     # reports
     color = fields.Char('Color', default="#875A7B")
 
+    _entry_limit_positive = models.Constraint(
```
That being said, unsure we need a sql constraint (implies some db processing that is not really important in our case). Simply managing anything that is not > 1 as "skip" works.

frva-odoo replied:
I removed the sql constraint but put a python constraint instead, because I think it would be weird as it does not make sense to have negative entry limit and could lead to errors later on if we let users input negative values in the field. If you still think this is overkill and useless, I'll remove it

tde-banana-odoo commented on addons/event/tests/test_event_internals.py:
```diff
@@ -907,6 +907,36 @@ def test_registration_partner_sync_company(self):
         self.assertEqual(new_reg.phone, contact.phone)
 
+    @users('user_eventmanager')
+    def test_registration_remaining_entries(self):
```
Should also cover "cancel_last_attendance"

tde-banana-odoo commented on addons/event/tests/test_event_internals.py:
```diff
@@ -907,6 +907,36 @@ def test_registration_partner_sync_company(self):
+        })
+        self.assertEqual(multi_entry_registration.remaining_entries, multi_entry_event_ticket.entry_limit)
+        multi_entry_registration.action_attend_event()
+        multi_entry_registration._compute_remaining_entries()
```
Invalidate cache instead probably

tde-banana-odoo commented on addons/event/tests/test_event_internals.py:
```diff
@@ -907,6 +907,36 @@ def test_registration_partner_sync_company(self):
+            'event_id': multi_entry_event.id,
+            'event_ticket_id': multi_entry_event_ticket.id,
+        })
+        self.assertEqual(multi_entry_registration.remaining_entries, multi_entry_event_ticket.entry_limit)
```
Would prefer comparing to exact value we expect (+ msg in case of fail)

tde-banana-odoo commented on addons/event/models/event_registration.py:
```diff
@@ -312,6 +329,48 @@ def action_set_draft(self):
+                new_record_fields = {
+                    'main_registration_id': record.id,
+                }
+                if self.env.context.get('mail_notrack'):
```
Why would calling "action_attend_event" with multi entries in form view send emails while only the barcode usage would not ? Is it really expected ?

frva-odoo replied:
I agree with you that this should change, one way or the other should not create a difference in the behavior like this.

We already discussed this in the task chatter but to me we shouldn't send any email from one way or the other, because that means that the attendee would receive an email with a used ticket whenever we scanned the main ticket that was bought in the first place. So he receives is ticket, goes to the event, scan it and receive a second ticket by mail (the sub registration) that is already validated which makes no sense to him. Same for the validation from the backend, the ticket would be already marked as used so it makes no sense. 

I think we should add in the event_mail _execute_attendee_based method the condition that registration should not have a main_registration_id. This would never send any mail for any sub-registration created that way. Tell me what you think about this or we can discuss it further if you want.

tde-banana-odoo replied:
Would agree on this, skip attendee-based mailing for sub registrations :) .

tde-banana-odoo replied:
That way we avoid the 'mail no track' field addition also.

frva-odoo commented:
Hello @tde-banana-odoo, thank you for your last review, I applied each comment and did a rebase so ready for you to review again :)

tde-banana-odoo commented on addons/event/models/event_registration.py:
```diff
@@ -78,6 +78,9 @@ def _get_random_barcode(self):
              'Cancelled: registrations cancelled manually')
+    remaining_entries = fields.Integer(string="Remaining Entries", compute="_compute_remaining_entries", tracking=True)
+    ticket_entry_limit = fields.Integer(string="Initial Entry Limit", related="event_ticket_id.entry_limit", tracking=True)
+    main_registration_id = fields.Many2one('event.registration', string='Main Registration', ondelete='restrict')
```
You can add a 'btree_not_null' index on this one, will be necessary I think

tde-banana-odoo commented on addons/event/models/event_registration.py:
```diff
@@ -312,6 +328,49 @@ def action_set_draft(self):
     def action_confirm(self):
         self.write({'state': 'open'})
 
+    def cancel_last_sub_registration(self):
```
:socks: Classic action naming (aka called by JS)
```suggestion
    def action_cancel_last_sub_registration(self):
```
Also to define after "action_cancel" to keep things organized :)

tde-banana-odoo commented on addons/event/models/event_registration.py:
```diff
@@ -312,6 +328,49 @@ def action_set_draft(self):
+            [reg_id for main_reg, reg_id in last_sub_registrations]
+        ).action_cancel()
+
+    def undo_registration_attendance(self):
```
:socks: Maybe, to keep naming state-based ? also to define after action_confirm
```suggestion
    def action_confirm_and_reset(self):
```

tde-banana-odoo commented on addons/event/models/event_registration.py:
```diff
@@ -312,6 +328,49 @@ def action_set_draft(self):
+                }
+                new_registration = record.copy(new_record_fields)
+                new_registration.action_set_done()
+                sub_registration_link = new_registration._get_html_link(title=f"{new_registration.name}")
```
:socks: Keep naming coherent
```suggestion
                new_registration_link = new_registration._get_html_link(title=f"{new_registration.name}")
```

tde-banana-odoo commented on addons/event/models/event_ticket.py:
```diff
@@ -135,6 +139,12 @@ def _constrains_limit_max_per_order(self):
+    @api.constrains('entry_limit')
+    def _constrains_entry_limit(self):
+        for ticket in self:
+            if ticket.entry_limit < 0:
```
Can be batched / one-lined
```suggestion
        failing = self.filtered(lambda t: t.entry_limit < 0)
```

tde-banana-odoo commented on addons/event/models/event_ticket.py:
```diff
@@ -135,6 +139,12 @@ def _constrains_limit_max_per_order(self):
+    def _constrains_entry_limit(self):
+        for ticket in self:
+            if ticket.entry_limit < 0:
+                raise UserError(_('The Entry Limit of an event ticket cannot be lower than 0.'))
```
Would include ticket name, in case of mass update, usefull to know which one failed
```suggestion
                raise UserError(_('The Entry Limit of event tickets %(ticket_names)s cannot be lower than 0.', ticket_names=', '.join(ticket.mapped('name'))))
```

tde-banana-odoo commented:
@robodoo r+ rebase-merge

Thanks for the followup !

robodoo commented:
Merge method set to rebase and merge, using the PR as merge commit message.
